protected/humhub/modules/tour/Events.php

```diff
@@ -0,0 +1,78 @@
+<?php
+/**
+ * @link https://www.humhub.org/
+ * @copyright Copyright (c) HumHub GmbH & Co. KG
+ * @license https://www.humhub.com/licences
+ */
+
+namespace humhub\modules\tour;
+
+use humhub\modules\dashboard\widgets\Sidebar;
+use humhub\modules\tour\widgets\Dashboard;
+use Yii;
+
+/**
+ * Events provides callbacks for all defined module events.
+ *
+ * @since 1.15
+ */
+class Events
+{
+    const AUTO_START_KEY = 'tour.autoStartWelcome';
+
+    public static function onDashboardSidebarInit($event)
+    {
+        if (Yii::$app->user->isGuest) {
+            return;
+        }
+
+        $settings = self::getModule()->settings;
+
+        if ($settings->get('enable') == 1 && $settings->user()->get('hideTourPanel') != 1) {
```

Comment:
Hmm, since we load the module user settings `$settings->user()->get('hideTourPanel')` here anyway, the optimization with the session is probably obsolete. :(

Reply:
@luke- If I understand you correctly you want a solution what I have implemented in new commit https://github.com/humhub/humhub/pull/6685/commits/23f4e22d8dd5e6e6ebba9b3abd97d82d4205717d, so new Session var(to display the dashboard tour widget) is initialized on the event `onAfterLogin`.

But please note the new Session flag will be updated only after re-login user, so I think here we may have 2 issues:
- for all logged in user the dashboard sidebar tour widget will be hidden after apply this update,
- the widget visibility will not be refreshed after change the settings `settings->get('enable')` and `settings->user()->get('hideTourPanel')`

Do you think it is normal for the tour widget? Or we should find a solution how to refresh it for these 2 cases?

Reply:
@yurabakhtin Not exactly, since we read out the user tour settings anyway, we can also replace the Login Event & Session solutionwith `$settings->user()->get('firstLoginPopupDisplayed')` here as well. 

However, we also need a solution so that the dialog does not appear for existing users.

Reply:
@luke- Sorry, not clear what to change.

Just for info: before this update https://github.com/humhub/humhub/pull/6249/files#diff-7151ee0d1890e8e7a3812a19e6b4f0c4bb287b1f485bdbd103d50f13b13c1dacL36-L38 the event `onDashboardSidebarInit` is used only for displaying of the tour widget on the dashboard sidebar.

I want to understand when we should use `$settings->user()->get('firstLoginPopupDisplayed')`:
- probably you mean to set this user setting if the tour window was forced to be displayed after login, because currently there code `$user->updated_by === null && $user->created_at === $user->updated_at` is used for this.
- or please describe your idea if what I wrote above is wrong, thanks.

Reply:
@yurabakhtin @luke- Sorry, I didn't fully thought fully of that myself.

How about the following:

1.) EVENT_AFTER_REGISTRATION sets `$tourSettings->user()->set('autoStartTour', 1)`
2.) onDashboardSidebarInit checks for `$tourSettings->user()->get('autoStartTour')` and starts tour and removes this setting.

Reply:
@luke- Ok, I have replaced the Session var to `$tourSettings->user()->get('autoStartTour')` in commit https://github.com/humhub/humhub/pull/6685/commits/a389af0dde8a8d60d313e00446c0f39a548f0afd:

- I have found the column `user.last_login` is NULL for new created/registered user,
- so instead of the updating `user.updated_by` as it was used as flag before, I have decided to compare `$user->last_login === null` on event `onUserBeforeLogin`,
- but we still use the Session var to display the tour widget in dashboard sidebar, this session var is refreshed only after each login action, I mean the same issue described here https://github.com/humhub/humhub/pull/6685#discussion_r1408926312, i.e. currently we have 2 flags:
  - Auto start the tour (initialized before login only for new registered users)
  - Display dashboard sidebar widget (initialized after login by condition `$settings->get('enable') == 1 && $settings->user($user)->get('hideTourPanel') != 1`, as I understood we don't need to call this on each dashboard sidebar init and it is why we cached it in the Session var).

Reply:
@luke- 

> Hmm, isn't `last_login` set on the first login? Sounds like a bug for me.

After registration or creating of new user the column `last_login` == `NULL`.
After each login it is filled with current date, it is why I have decided to use the column as flag of first logging even after user creating by admin and not only after registration (`Registration::EVENT_AFTER_REGISTRATION`)

> Could an intermediate Term&Condition page possibly interfere here?

I think no, I have tested this case right now, so current implementaion works this way:

- New user has `last_login = NULL`
- User is logged in first time
- Event `User::EVENT_BEFORE_LOGIN` is called, there we set `user->settings->set('autoStartTour', true)`
- Redirect to page "Terms and Conditions", two case:
  - Logout => the user setting `autoStartTour` is stored for next login action.
  - Accept => go do bashboard, if the dashboard sidebar widget is visible so event `onDashboardSidebarInit` is called:
    - if `autoStartTour === true` => run the auto tour
    - else display the Tour Widget in the Dashboard Sidebar, but I don't like that we use here the condition `Yii::$app->session->get(self::DASHBOARD_WIDGET)` to display it, this Session var is set on the event `AuthController::EVENT_AFTER_LOGIN`, i.e. the Tour Widget visibility will be updated per User only after each relogin, if you think it is ok, then we can keep this solution, otherwise we should improve it to refresh the Session somehow after update the settings which are used to set the Session var `self::DASHBOARD_WIDGET`.

Reply:
> I'm not sure if I'm missing something here, but wouldn't following approach the easiest solution:
> 
> **Event Handler `Registration::EVENT_AFTER_REGISTRATION`:**
> 
> ```
>         /** @var \humhub\modules\tour\Module $tour */
>         $tour = Yii::$app->getModule('tour');
>         $tour->settings->user($event->identity)->set('doAutostart', 1);
> ```

@luke- As I wrote above I prefer to use event `User::EVENT_BEFORE_LOGIN` together with flag `last_login === NULL`, in order to don't skip case when user is creating from Administration or API.

> **Event Tour Widget:**
> 
> ```
>     public static function onDashboardSidebarInit($event)
>     {
>           /** @var Module $module */
>         $module = Yii::$app->getModule('tour');
> 
>         $doAutoStart = false;
>         if (!empty($module->settings->user()->get('doAutostart'))) {
>             $module->settings->user()->delete('doAutostart');
>             $doAutostart = true;
>         }
>         
>         // ... pass $doAutostart to the widget
>     }
> ```

Yes, it works this way.

> Regarding Session Cache: I don't think this is useful, because `$settings->user($user)->get('hide')` will load all user settings (incl. autoStart) for the tour module anyway.

Currently we store in Session the condition to display the Tour Widget on the Dashboard Sidebar:

![tour-widget](https://github.com/humhub/humhub/assets/6995524/8892c7f8-e3f4-4d70-b555-ac57703b2737)

I have done to store it in Session because as I understood here https://github.com/humhub/humhub/pull/6685#discussion_r1408444095 you don't like to use there data from database.
But I don't think it is good solution to use Session var there, so I have reverted that and prefer to use it as I have done now in the commit https://github.com/humhub/humhub/pull/6685/commits/4ecaad9b7e5a269518cf20552391d77afe0df0ac, do you agree this or please tell why we cannnot this solution?